runtime/src/iree/hal/drivers/amdgpu/util/libhsa.c

```diff
@@ -0,0 +1,692 @@
+      return "The arguments passed to a functions are not compatible.";
+    case HSA_STATUS_ERROR_INVALID_INDEX:
+      return "The index is invalid.";
+    case HSA_STATUS_ERROR_INVALID_ISA:
+      return "The instruction set architecture is invalid.";
+    case HSA_STATUS_ERROR_INVALID_ISA_NAME:
+      return "The instruction set architecture name is invalid.";
+    case HSA_STATUS_ERROR_INVALID_CODE_OBJECT:
+      return "The code object is invalid.";
+    case HSA_STATUS_ERROR_INVALID_EXECUTABLE:
+      return "The executable is invalid.";
+    case HSA_STATUS_ERROR_FROZEN_EXECUTABLE:
+      return "The executable is frozen.";
+    case HSA_STATUS_ERROR_INVALID_SYMBOL_NAME:
+      return "There is no symbol with the given name.";
+    case HSA_STATUS_ERROR_VARIABLE_ALREADY_DEFINED:
+      return "The variable is already defined.";
+    case HSA_STATUS_ERROR_VARIABLE_UNDEFINED:
+      return "The variable is undefined.";
+    case HSA_STATUS_ERROR_EXCEPTION:
+      return "An HSAIL operation resulted in a hardware exception.";
+    case HSA_STATUS_ERROR_INVALID_CODE_SYMBOL:
+      return "The code object symbol is invalid.";
+    case HSA_STATUS_ERROR_INVALID_EXECUTABLE_SYMBOL:
+      return "The executable symbol is invalid.";
+    case HSA_STATUS_ERROR_INVALID_FILE:
+      return "The file descriptor is invalid.";
+    case HSA_STATUS_ERROR_INVALID_CODE_OBJECT_READER:
+      return "The code object reader is invalid.";
+    case HSA_STATUS_ERROR_INVALID_CACHE:
+      return "The cache is invalid.";
+    case HSA_STATUS_ERROR_INVALID_WAVEFRONT:
+      return "The wavefront is invalid.";
+    case HSA_STATUS_ERROR_INVALID_SIGNAL_GROUP:
+      return "The signal group is invalid.";
+    case HSA_STATUS_ERROR_INVALID_RUNTIME_STATE:
+      return "The HSA runtime is not in the configuration state.";
+    case HSA_STATUS_ERROR_FATAL:
+      return "The queue received an error that may require process "
+             "termination.";
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// iree_hal_amdgpu_libhsa_t
+//===----------------------------------------------------------------------===//
+
+#if !IREE_HAL_AMDGPU_LIBHSA_STATIC
+
+static iree_status_t iree_hal_amdgpu_libhsa_load_symbols(
+    iree_dynamic_library_t* library, iree_hal_amdgpu_libhsa_t* out_libhsa) {
+#define IREE_HAL_AMDGPU_LIBHSA_PFN(trace_category, result_type, symbol, ...) \
+  IREE_RETURN_IF_ERROR(iree_dynamic_library_lookup_symbol(                   \
+      library, #symbol, (void**)&out_libhsa->symbol));
+#include "iree/hal/drivers/amdgpu/util/libhsa_tables.h"  // IWYU pragma: keep
+  return iree_ok_status();
+}
+
+// TODO(benvanik): move someplace central - status and string_builder have a
+// cycle and it's tricky to move there right now.
+static bool iree_string_builder_append_status(iree_string_builder_t* builder,
+                                              iree_status_t status) {
```

Comment:
+1 for the TODO - weird having a function in here that isn't in the `iree_hal_amdgpu_` namespace, and other parts of the project may want to use this function without redefining it too.